Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,424 @@
+it will slow that process down. And if you rebuild an index then any slow code running in these events will cause the indexing
+to go ultra slow.
+
+## RenderTemplateAsync
+
+There is an API in Umbraco that should never be used unless you really know what you are doing. This API method
+is called `RenderTemplateAsync`. It allows you to be able to render a particular content item's template and get a `IHtmlEncodedString`
+in response. In some cases, this may be useful. Perhaps you want to send an email based on a content item and its template, but
+you must be very careful not to use this for purposes it is not meant to be used for.
+
+Generally speaking, this method should not be used for the normal rendering of content. If abused this could cause severe
+performance problems. For normal content rendering of module type data from another content item, you should use Partial Views instead.
+
+## Don't put logic inside your constructors
+
+Constructors should generally not perform any logic, they should set some parameter values, perform some null checks and perhaps validate some data but in most cases, they should not perform any logic.
+
+There are a few reasons why this can become a huge performance problem:
+
+* The consumer of an API doesn't expect that by creating an object that they should be worried about performance
+* Creating an object can inadvertently happen a vast number of times, especially when using LINQ
+
+Here's an example of how this can go wrong very quickly:
+Your tree structure is something like this:
+
+    - Root
+    -- Home
+    --- Recipes (node id = 3251, list view with 5000 nodes)
+    --- About Us
+    --- Contact Us
+
+You have a custom model that looks like this:
+
+```csharp
+public class RecipeModel : PublishedContentWrapped
+{
+    public RecipeModel(IPublishedContent content, IPublishedValueFallback publishedValueFallback) : base(content, publishedValueFallback)
+    {
+        RelatedRecipes = content
+            .Parent
+            .Children<RecipeModel>()
+            .Where(x => x.Value<IEnumerable<int>>("related")
+                .Contains(content.Id));
+
+        Votes = content.Value<int>("votes");
+    }
+    
+    public int Votes { get; private set; }
+    
+    public IEnumerable<RecipeModel> RelatedRecipes { get; private set; }
+}
+}
+```
+
+You then run the following code to show to show the favorites
+
+```csharp
+@var recipeNode = Umbraco.TypedContent(3251);
+@{
+	var recipeNode = Umbraco.Content(1234);
+}
+
+<ul>
+	@foreach (var recipe in recipeNode.Children
+		.Select(x => new RecipeModel(x, _publishedValueFallback))
+		.OrderByDescending(x => x.Votes)
+		.Take(10))
+	{
+		<li><a href="@recipe.Url()">@recipe.Name</a></li>
+	}	
+</ul>
+
+```
+
+__Ouch!__ To show the top 10 voted recipes's this will end up doing the following:
+
+* This will iterate over all Recipes, create and allocate 5000 instances of `IPublishedContent`
+* This will create and allocate 5000 instances of `RecipeModel`
```

Comment:
Since the example does `.Select(x => new RecipeModel(x, _publishedValueFallback))`, then I suppose we'll still create an allocate 5000 instances of `RecipeModel` even though the initial content is already in the cache? Since the Linq needs to create the objects to order by votes, so I've left that in, but adjusted the rest accordingly :D